p2p/src/cybershake.rs

```diff
@@ -286,6 +378,121 @@ impl<W: AsyncRead + Unpin> Incoming<W> {
     }
 }
 
+impl<W: AsyncRead + Unpin> Incoming<W> {
+    fn decipher_buf(&mut self) {
+        let seq = self.seq;
+        self.seq += 1;
+
+        self.kdf.append_u64(b"seq", seq);
+        let mut key = [0u8; 32];
+        self.kdf.challenge_bytes(b"key", &mut key);
+
+        let ad = encode_u64le(seq);
+
+        let siv_tag = GenericArray::clone_from_slice(&self.buf.as_slice()[..16]);
+        Aes128PmacSiv::new(GenericArray::clone_from_slice(&key))
+            .decrypt_in_place_detached(
+                &[&ad],
+                &mut self.buf.as_mut_slice()[16..self.ciphertext_length as usize],
+                &siv_tag,
+            )
+            .unwrap();
+
+        let pt_len = self.ciphertext_length as usize - 16;
+
+        for i in 0..pt_len {
+            let byte = self.buf.as_slice()[i + 16];
+            self.buf.as_mut_slice()[i] = byte;
+        }
+
+        self.plaintext_length = pt_len;
+        self.ciphertext_length = 0;
+        self.ciphertext_read = 0;
+    }
+
+    fn read_buffer(&mut self, buf: &mut [u8]) -> Result<usize, io::Error> {
+        match Read::read(
+            &mut &self.buf[self.plaintext_read..self.plaintext_length],
```

Comment:
Can we please not use the Read trait here? It is a mutable trait that's supposed to update offset of the backend, but we are doing it manually here via the `self.plaintext_read` offset. Passing a mutable reference to our buffer makes me wonder what's going to be changed. It'd be much clearer if we checked for conditions separately:

1. If the requested buf is longer than what we have left - read till the end and return that much.
2. If it's shorter or equal - read as much as requested.

Reply:
Fixed